src/main/kotlin/org/rust/ide/typing/paste/JsonStructParser.kt

```diff
@@ -0,0 +1,140 @@
+/*
+ * Use of this source code is governed by the MIT license that can be
+ * found in the LICENSE file.
+ */
+
+package org.rust.ide.typing.paste
+
+import com.fasterxml.jackson.core.JsonFactory
+import com.fasterxml.jackson.core.JsonParser
+import com.fasterxml.jackson.core.JsonToken
+import java.io.IOException
+
+sealed class DataType {
+    object Integer : DataType()
+    object Float : DataType()
+    object Boolean : DataType()
+    object String : DataType()
+    object Unknown : DataType()
+    object Missing : DataType()
+    data class StructRef(val struct: Struct) : DataType()
+    data class Array(val type: DataType) : DataType()
+}
+
+data class Field(val key: String, val type: DataType)
+
+class Struct(val fields: List<Field>) {
+    // Consider structs with the same field names and types to be the same
+    // The fields are sorted before comparison, so two struct with the same field types and names, but
+    // with a different field order, are considered to be the same struct.
+    override fun equals(other: Any?): Boolean {
+        if (other == null) return false
+        if (other !is Struct) return false
+
+        val selfFields = fields.sortedBy { it.key }
+        val otherFields = other.fields.sortedBy { it.key }
+        return selfFields == otherFields
+    }
+
+    override fun hashCode(): Int = fields.sortedBy { it.key }.hashCode()
+}
+
+/**
+ * Try to parse the input text as a JSON object.
+ *
+ * Extract a list of (unique) structs that were encountered in the JSON object.
+ */
+fun extractStructsFromJson(text: String): List<Struct>? {
+    return try {
+        val factory = JsonFactory()
+        val parser = factory.createParser(text)
+
+        val structParser = StructParser()
+        parser.use {
+            // There must be an object at the root
+            val token = parser.nextToken()
+            if (token != JsonToken.START_OBJECT) return null
+
+            structParser.parseStruct(parser)
+        }
+
+        structParser.encounteredStructs
+    } catch (e: IOException) {
+        null
+    }
+}
+
+private class StructParser {
+    private val structMap = mutableSetOf<Struct>()
+    private val structs = mutableListOf<Struct>()
+
+    val encounteredStructs: List<Struct> = structs
```

Comment:
```suggestion
    private val structs = linkedSetOf<Struct>()

    val encounteredStructs: List<Struct> get() = structs.toList()
```

Please don't use `mutableSetOf`/`mutableMapOf`. Use explicit `linkedSetOf` if ordering is needed and `hashSetOf` otherwise (for slightly better memory consumption)